### ISSUE TITLE
[Truffle] Incorrect string size

### ISSUE THREAD

**Issue by bjfish** [jruby/jruby#4396]
I've ran into this issue while working on encode specs / scrub stuff

### Example
```
x81 = [0x81].pack('C').force_encoding('utf-8')
puts "abc\u3042#{x81}".size
```

### Expected Behavior

```
$ ruby ../jruby-patches/str_size.rb 
5
```

### Actual Behavior
```
$ jt run ../jruby-patches/str_size.rb 
7
```

**Comment by nirvdrum** [comment]
I'll investigate more. At first blush, you're probably hitting this TODO: https://github.com/jruby/jruby/blob/f17102c9d9fcfab3108bfb32c4d349f1d4ba2998/truffle/src/main/java/org/jruby/truffle/core/rope/ValidLeafRope.java#L27-L28

**Comment by nirvdrum** [comment]
I've fixed this in 3f4af191bed1c5077527afecdcac7f32af264411

The real problem is we just reported the byte length as the string length for broken strings while MRI tries to calculate the character length and stops when it encounters an invalid byte sequence. Neither value is correct -- the string is broken so you can't calculate the correct value. So, I previously opted to use a faster calculation. But, broken strings are likely to be infrequent so we can be slower to match MRI. However, I also wouldn't recommend relying on that value since it seems like something that's not well defined.
